scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala

```diff
@@ -296,6 +296,29 @@ object TreeSyntax {
       }
       !isOnlyChildOfOnlyChild(t)
     }
+    @annotation.tailrec
+    def guessNeedsLineSep(t: Tree): Boolean = t match {
+      case _: Decl.Val | _: Decl.Var | _: Decl.Def | _: Defn.Macro | _: Defn.Type | _: Term.This |
```

Comment:
why do you include `Decl` types here? they can't be confused with definitions because they all require an `=` before the body, so a block after shouldn't be confused. (as mentioned below, that includes the procedure syntax because a declaration must have a type and procedure syntax doesn't.)

if I am mistaken about that, there are 5 `Decl` classes, you didn't include `Given` and `Type`. why? should we match simply on `_: Decl`, to simplify matters (just like with `Tree.WithBody`)?

Reply:
The Decl.{Val, Var, Def} require the line separator, while Decl.Type does not. I don't know the reason but the Scala 2.13 parser fails on those. The tests that I created check that the line separator is actually necessary (because it either fails to parse without it or because it changes the tree structure).

The Decl.Given AFAIK is Scala3 only so it should never reach the `guessNeedsLineSep` function. In the call site we are checking for `dialect.allowSignificantIndentation`:
`if (dialect.allowSignificantIndentation) x.map(i(_)) else withBlankLines(x)`

Maybe we should add the Decl.Given (and other scala3 only Stats) anyways for future-proof? Is there an easy way to parse Scala3 without this `allowSignificantIndentation` flag on? If this makes sense then I'll have to brush up on my Scala 3 syntax, I don't use it on my day-to-day job :)

Reply:
in that case, let's leave it as-is.